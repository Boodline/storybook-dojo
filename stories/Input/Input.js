import * as React from 'react'
import PropTypes from 'prop-types';

export const Input = (
  { primary, backgroundColor, size, ...props }
) => {

  const { 
    id,
    value,
    onChangeValue,
    type,
    label,
    required,
    placeholder,
    disable
   } = props

  const [localValue, changeLocalValue] = React.useState(value)

  const onChangeAction = (event) => {
    const newValue = event.currentTarget.value
    changeLocalValue(newValue)
    onChangeValue(id, newValue)
  }

  React.useEffect(() => {
    if (value !== localValue) {
      changeLocalValue(value)
    }
  }, [value])

  return (
    <div >
      <label>{`${label}${required ? '*' : ''}`}</label>
      <div>
        <input
          disabled={disable}
          id={id}
          value={localValue}
          onChange={onChangeAction}
          type={type}
          placeholder={placeholder}
          style={backgroundColor && { backgroundColor }}
        />
      </div>
    </div>
  )
}

Input.propTypes = {
  id: PropTypes.string,
  value: PropTypes.string,
  onChangeValue: PropTypes.func,
  type: PropTypes.oneOf(['text', 'email', 'number', 'password']),
  label: PropTypes.string,
  required: PropTypes.bool,
  placeholder: PropTypes.string,
  disable: PropTypes.bool,
}

Input.defaultProps = {
  id: 'input-component',
  value: '',
  onChangeValue: () => null,
  type: 'text',
  label: 'Label',
  required: false,
  placeholder: 'write here',
  disable: false,
};
