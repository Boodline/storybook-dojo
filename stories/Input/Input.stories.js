import React from 'react';
import { Input } from './Input';
import { action } from '@storybook/addon-actions'

export default {
  title: 'Example/Input',
  component: Input,
  argTypes: {
    backgroundColor: { control: 'color' },
  },
};

const Template = (args) => <Input {...args} />;

export const Basic = Template.bind({});
Basic.args = {
  id: 'input-component',
  value: '',
  onChangeValue: action('changeValue'),
  type: 'text',
  label: 'Label',
  required: false,
  placeholder: 'write here',
  disable: false,
};
