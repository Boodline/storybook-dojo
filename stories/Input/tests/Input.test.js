import * as renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react'
import { Input } from '../Input'
import * as React from 'react'


describe('Input component tests', () => {
  it('should render component', () => {
    const inputComponent = <Input
      id='input-component'
      value=''
      onChangeValue={null}
      type='text'
      label='Label'
      required={false}
      placeholder='write here'
      disable={false}
    />

    const component = renderer.create(inputComponent)
    const componentJSON = component.toJSON()
    expect(componentJSON).toMatchSnapshot()
  })

  it('should execute sent function on keyTyped', () => {
    const act = jest.fn()
    const { getByTestId } = render(<Input
      id='input-component'
      value=''
      onChangeValue={act}
      type='text'
      label='Label'
      required={false}
      placeholder='write here'
      disable={false}
    />)

    renderer.act(() => {
      fireEvent.change(getByTestId('input-component'), { target: { value: 'Hello World!' }})
    })
    expect(act).toBeCalledTimes(1)
  })

  it('should change component value', () => {
    const act = jest.fn()
    const { getByTestId } = render(<Input
      id='input-component'
      value=''
      onChangeValue={act}
      type='text'
      label='Label'
      required={false}
      placeholder='write here'
      disable={false}
    />)

    renderer.act(() => {
      fireEvent.change(getByTestId('input-component'), { target: { value: 'Hello World!' }})
    })
    expect(getByTestId('input-component').getAttribute('value')).toBe('Hello World!')
  })

})