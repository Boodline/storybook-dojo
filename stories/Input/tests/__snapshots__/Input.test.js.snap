// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component tests should render component 1`] = `
<div>
  <label>
    Label
  </label>
  <div>
    <input
      data-testid="input-component"
      disabled={false}
      id="input-component"
      onChange={[Function]}
      placeholder="write here"
      type="text"
      value=""
    />
  </div>
</div>
`;
